### DISCUSSION TITLE
"by_alias" default: FastAPI vs Pydantic

### DISCUSSION THREAD

**Discussion by tomsem** [fastapi/fastapi#2753] [Questions] [answered]
In Pydantic the ```by_alias``` property controls the serialization of Pydantic objects, e.g.:
```
>>> from pydantic import BaseModel, Field
>>> class SomeClass(BaseModel):
...    some_variable: str = Field(alias='someVariable')
...    
>>> some_obj = SomeClass(**{'someVariable': 'some value'})
>>> some_obj.dict()
{'some_variable': 'some value'}
>>> some_obj.dict(by_alias=True)
{'someVariable': 'some value'}
```
The default is False, i.e. 'use the class variable name'.

However, when FastAPI serializes a Pydantic object in Response, it sets this property to True by default, and exports the alias name.

I can see an argument for either default, but it was 'surprising' to discover that they were different.

After being surprised, I discovered ```response_model_by_alias``` property that can be put in router definition to override the FastAPI default.

**Comment by syedhabib53** [comment]
Yes, But `response_model_by_alias` wont work with [beanie](https://beanie-odm.dev), since it uses alias name to map mongodb's primary key `_id` to `id`. If we use `response_model` for response, it wont work. Since `serialize_response` doesn't use `response_model_by_alias` [here](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L77). By default it uses `by_alias=True`. It only used in `jsonable_encoder` to encode the response. 

Right now I am using/calling every time `.dict()` method to avoid this validation error 😔.
```
response -> id
  field required (type=value_error.missing)
```

**Comment by tiangolo** [comment, marked as answer]
Yep, because if you define an alias, you would normally want it used in the API. And as you are not serializing it yourself, FastAPI has to try and have the most "sensible" default.

**Comment by seandlg** [comment]
It seems like this was changed recently? I updated `FastaAPI` to `0.89.1` and the following:

```python
from fastapi import FastAPI, Path, Query
from pydantic import BaseModel, Field

app = FastAPI()

class Item(BaseModel):
    password: str = Field(..., alias="pass")

@app.post("/item")
async def read_items(
    item: Item,
):
    print(item.dict())
    print(item.dict(by_alias=False))
    print(item.dict(by_alias=True))
    return item.dict(by_alias=True)```
```

prints:

```txt
{'password': 'string'}
{'password': 'string'}
{'pass': 'string'}
```

Notably, this differs from the behavior I was seeing previously. Here the `alias` was used by default. Will add the `Pydantic` version here as soon as I get access to the source.
